PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -699,33 +1215,14 @@ void ProjectPanel::setWorkSpaceDirty(bool isDirty)
 	_treeView.setItemImage(_treeView.getRoot(), iImg, iImg);
 }
 
-NodeType ProjectPanel::getNodeType(HTREEITEM hItem)
+ProjectPanel::NodeType ProjectPanel::getNodeType(HTREEITEM hItem) const
 {
 	TVITEM tvItem;
 	tvItem.hItem = hItem;
-	tvItem.mask = TVIF_IMAGE | TVIF_PARAM;
-	SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0,(LPARAM)&tvItem);
-
-	// Root
-	if (tvItem.iImage == INDEX_CLEAN_ROOT || tvItem.iImage == INDEX_DIRTY_ROOT)
-	{
-		return nodeType_root;
-	}
-	// Project
-	else if (tvItem.iImage == INDEX_PROJECT)
-	{
-		return nodeType_project;
-	}
-	// Folder
-	else if (tvItem.lParam == NULL)
-	{
-		return nodeType_folder;
-	}
-	// File
-	else
-	{
-		return nodeType_file;
-	}
+	tvItem.mask = TVIF_PARAM;
+	SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.